lib/puppet/util/windows/process.rb

```diff
@@ -224,6 +224,56 @@ def windows_major_version
   end
   module_function :windows_major_version
 
+  ENVSTRINGS_TERMINATOR_WCHAR = [0,0]
+  # Returns a hash of the current environment variables encoded as UTF-8
+  # The memory block returned from GetEnvironmentStringsW is double-null terminated and the vars are paired as below;
+  # Var1=Value1\0
+  # Var2=Value2\0
+  # VarX=ValueX\0\0
+  # Note - Some env variable names start with '=' and are safe to ignore
+  # Note - The env_ptr MUST be freed using the FreeEnvironmentStringsW function
+  def get_environment_strings
+    env_ptr = GetEnvironmentStringsW()
+
+    env_hash = {}
+    env_ptr.read_arbitrary_wide_string_up_to(4096, ENVSTRINGS_TERMINATOR_WCHAR).split(?\x00).each do |env_pair|
+      pair = env_pair.split("=")
+      unless pair[0] == ""
+        if (pair[1].nil?)
+          # Don't store nils in the hash otherwise the variable will be removed if called with set_environment_variable
+          env_hash.store(pair[0],"")
+        else
+          env_hash.store(pair[0],pair[1])
+        end
+      end
+    end
+    env_hash
+  ensure
+    if (!env_ptr.null?)
+      if FreeEnvironmentStringsW(env_ptr) == FFI::WIN32_FALSE
+        raise Puppet::Util::Windows::Error.new("FreeEnvironmentStringsW failed")
+      end
+    end
+  end
+  module_function :get_environment_strings
+
+  def set_environment_variable(name,val)
+    FFI::MemoryPointer.from_string_to_wide_string(name) do |name_ptr|
+      if (val.nil?)
```

Comment:
Might want to consider collapsing the logic / branching here... or adding a separate `remove_environment_variable(name)` helper that gets called when `val` is `nil` or `''`